# Ternas de primos potenciados
## Consigna

El 28 es el menor número que se puede expresar como ```a^2+b^3+c^4``` siendo a,b y c números primos.

```28 = 2^2+2^3+2^4
33 = 3^2+2^3+2^4
49 = 5^2+2^3+2^4
47 = 2^2+3^3+2^4
```

Contar todos los números menores de 50 millones que pueden ser expresados de este modo. 

*TIP: Hacer uso de la claúsula **break** para reducir la cantidad de iteraciones innecesarias.*

## Resolución

Para comenzar con este problema, debemos conocer los número primos. Pero la pregunta es cuantos números priomos necesitamos. 

Tomemos el término a^2. Si calculamos la raiz cuadrada de 50M entonces ese será el número primo más grande que deberemos calcular, ya que cualquier otro ya el número final sería más grande que nuestro límite.

Comencemos con definir el límite y calcular su raiz.

In [1]:
LIMIT=50000000
LIMIT**(1/2)

7071.067811865475

Ahora sabemos que cualquier número primo mayor a 7071 sobrepasa a LIMIT. Vamos a calcular entonces todos los números primos menores o iguales a 7071

In [2]:
def is_prime( n ):
    d = 2
    while d<n:
        if n%d==0: return False
        d+=1
    return True 

In [3]:
primes = list()
n=2
while n < 7072:
    if is_prime(n):
        primes += [n] 
    n += 1

Verificamos los primeros primos de la lista:

In [5]:
primes[0:10]

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

Ahora sólo nos queda iterar sobre estas posibilidades y encontrar todos los números resultantes. Para guardar los resultados, vamos a usar un set, que nos elimina los repetidos.

Dado que vamos a ir probando llos números de menor a mayor, si el ciclo ya vemos que superamos el LIMIT entonces ya no tiene sentido seguir probando, así que vamos a usar el **break** para salir de ese ciclo y acortar los tiempos de búsqueda.

Al final sólo contaremos la cantidad de elementos en el set.

In [7]:
nums = set()
for i in range(len(primes)):
    for j in range(len(primes)):
        if(primes[i]**2 + primes[j]**3)>= LIMIT: break ## Continuo con el siguiente i
        for k in range(len(primes)):
            n = primes[i]**2 + primes[j]**3 + primes[k]**4
            if n < LIMIT:
                nums.add(n)
            else:
                break ## Continuo con el siguiente j
print('Cantidad de números menores a',LIMIT,':',len(nums))

Cantidad de números menores a 50000000 : 1097343
